core/worker.go

```diff
@@ -871,20 +965,40 @@ func (w *worker) commitUncle(env *environment, uncle *types.WorkObjectHeader) er
 	if _, exist := env.uncles[hash]; exist {
 		return errors.New("uncle not unique")
 	}
-	var workShare bool
-	// If the uncle is a workshare, we should allow siblings
-	_, err := w.engine.VerifySeal(uncle)
-	if err != nil {
-		workShare = true
-	}
-	_, err = w.hc.WorkShareDistance(env.wo, uncle)
+
+	_, err := w.hc.WorkShareDistance(env.wo, uncle)
 	if err != nil {
+
+		if errors.Is(err, errInvalidWorkShareDist) {
+			// If uncle is found to be invalid because of distance, remove it from
+			// the share cache
+			var powid types.PowID
+			if uncle.AuxPow() == nil {
+				powid = types.Progpow
+			} else {
+				powid = uncle.AuxPow().PowID()
+			}
+
+			switch powid {
+			case types.Progpow, types.Kawpow:
+				w.kawpowShares.Remove(uncle.Hash())
+			case types.SHA_BCH, types.SHA_BTC:
+				w.shaShares.Remove(uncle.Hash())
+			case types.Scrypt:
+				w.scryptShares.Remove(uncle.Hash())
+			}
+		}
+
 		return err
 	}
-	if uncle.PrimaryCoinbase().IsInQiLedgerScope() && env.wo.PrimeTerminusNumber().Uint64() < params.ControllerKickInBlock {
-		return errors.New("workshare coinbase is in Qi, but Qi is disabled")
+	if env.wo.PrimeTerminusNumber().Uint64() < params.KawPowForkBlock || uncle.AuxPow() == nil || uncle.AuxPow().PowID() == types.Kawpow {
```

Comment:
Is this the right conditional? Specificall || PowID ==  types.Kawpow?

Reply:
technically it doesnt matter as the fork is happening after the controllerkickin block but we need consistency